src/OpenCL.cpp

```diff
@@ -260,6 +258,150 @@ __kernel void out_transform_fused_bn(__global float *M,
+                                     __local float *ybuf) {
+    const int W = 19;
+    const int H = 19;
+    const int T = W*H;
+    const int WTILES = (W + 1) / 2;
+    const int P = WTILES * WTILES;
+    const int KPpad = Kpad * Ppad;
+
+    const int k = get_global_id(0);
+    const int kg = get_local_id(0);
+    const int block = get_global_id(1);
+    const int chT = k*(T);
+
+    const int block_x = block % WTILES;
+    const int block_y = block / WTILES;
+
+    const int yin = 2 * block_y - 1;
+    const int xin = 2 * block_x - 1;
+
+
+    const int x = 2*block_x;
+    const int y = 2*block_y;
+    int a_ind = (y)*W + (x);
+
+    const int kHW = k * W * H;
+
+    if (k < K && block < P) {
+        const int a[4] = {a_ind, a_ind+1, a_ind+W, a_ind+W+1};
+        const bool pred[4] = { 1, x+1 < W, y+1 < H, x+1 < W & y+1 < H};
+        int b = block_y * WTILES + block_x;
+        float temp_m[16];
+        for (int xn = 0, xnKPpad = b*Kpad + k; xn < 16; xn++, xnKPpad += KPpad) {
+            temp_m[xn] = M[xnKPpad];
+        }
+
+        float o[4];
+        o[0] = temp_m[0*4 + 0] + temp_m[0*4 + 1] + temp_m[0*4 + 2] +
+               temp_m[1*4 + 0] + temp_m[1*4 + 1] + temp_m[1*4 + 2] +
+               temp_m[2*4 + 0] + temp_m[2*4 + 1] + temp_m[2*4 + 2];
+
+        o[1] = temp_m[0*4 + 1] - temp_m[0*4 + 2] - temp_m[0*4 + 3] +
+               temp_m[1*4 + 1] - temp_m[1*4 + 2] - temp_m[1*4 + 3] +
+               temp_m[2*4 + 1] - temp_m[2*4 + 2] - temp_m[2*4 + 3];
+
+        o[2] = temp_m[1*4 + 0] + temp_m[1*4 + 1] + temp_m[1*4 + 2] -
+               temp_m[2*4 + 0] - temp_m[2*4 + 1] - temp_m[2*4 + 2] -
+               temp_m[3*4 + 0] - temp_m[3*4 + 1] - temp_m[3*4 + 2];
+
+        o[3] = temp_m[1*4 + 1] - temp_m[1*4 + 2] - temp_m[1*4 + 3] -
+               temp_m[2*4 + 1] + temp_m[2*4 + 2] + temp_m[2*4 + 3] -
+               temp_m[3*4 + 1] + temp_m[3*4 + 2] + temp_m[3*4 + 3];
+
+        const float mean = vload_net_t(k, means);
+        const float scale_stddiv = vload_net_t(k, stddivs);
+
+        for (int i = 0; i < 4; i++) {
+            if (pred[i]) {
+                o[i] = scale_stddiv * (o[i] - mean);
+                if (residual) {
+                    o[i] += vload_net_t(kHW + a[i], residual);
+                }
+                o[i] = o[i] > 0 ? o[i] : 0.0f;
+                ybuf[kg * T + a[i]] = o[i];
+                if (Y) {
+                    vstore_net_t(o[i], kHW + a[i], Y);
+                }
+            }
+        }
+    }
+
+    barrier(CLK_LOCAL_MEM_FENCE);
+
+    if (block < P && k < K) {
+        const int CPpad = Ppad * Cpad;
+        // Cache input tile and handle zero padding
+        float xx[4][4];
+        for (int i = 0; i < 4; i++) {
+            for (int j = 0; j < 4; j++) {
+                int a = xin + j;
+                int b = yin + i;
+                if (b >= 0 && a >= 0 && b < H && a < W) {
+                    xx[i][j] = ybuf[kg * T + b*W + a];
```

Comment:
So from what I can tell, up to 16 locations are read from ybuf[] but up on line 333 only up to 4 locations are written to. Are the other locations initialised elsewhere? Or should the code here be calling vload_net_t() for those that weren't loaded into ybuf[]?

Reply:
OK, I just read what barrier(CLK_LOCAL_MEM_FENCE) does :)